reasoner/planner/AnswerCountEstimator.java

```diff
@@ -60,210 +59,204 @@
+            // TODO: ^ Does it though?
+            AnswerCountEstimator.ConjunctionModel acyclicModel = conjunctionModelFactory.buildAcyclicModel(conjunctionSummary);
             conjunctionModels.put(conjunction, acyclicModel);
 
             // cyclic calls to this model will answer based on the acyclic model.
-            iterate(conjunctionContext.cyclicConcludables()).flatMap(concludable -> iterate(dependencies(concludable)))
+            iterate(conjunctionSummary.cyclicConcludables()).flatMap(concludable -> iterate(conjunctionSummarizer.dependencies(concludable)))
                     .forEachRemaining(this::buildConjunctionModel);
-            ConjunctionModel cyclicModel = conjunctionModelFactory.buildCyclicModel(conjunctionContext, acyclicModel);
+            AnswerCountEstimator.ConjunctionModel cyclicModel = conjunctionModelFactory.buildCyclicModel(conjunctionSummary, acyclicModel);
             conjunctionModels.put(conjunction, cyclicModel);
         }
     }
 
-    private Set<ResolvableConjunction> dependencies(Concludable concludable) {
-        return iterate(logicMgr.applicableRules(concludable).keySet()).map(rule -> rule.condition().conjunction()).toSet();
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variables) {
+        if (!conjunctionModels.containsKey(conjunction)) buildConjunctionModel(conjunction);
+        if (!fullConjunctionEstimators.containsKey(conjunction)) {
+            IncrementalEstimator incrementalEstimator = createIncrementalEstimator(conjunction);
+            conjunctionModels.get(conjunction).conjunctionSummary.resolvables().forEach(incrementalEstimator::extend);
+            fullConjunctionEstimators.put(conjunction, incrementalEstimator);
+        }
+        IncrementalEstimator estimator = fullConjunctionEstimators.get(conjunction);
+        Set<Variable> estimateableVariables = ReasonerPlanner.estimateableVariables(variables);
+        return estimator.answerEstimate(estimateableVariables);
     }
 
-    private static class ConjunctionConcludableStack {
-
-        ArrayList<ResolvableConjunction> stack;
-        Map<ResolvableConjunction, Concludable> concludables;
-
-        ConjunctionConcludableStack() {
-            this.stack = new ArrayList<>();
-            this.concludables = new HashMap<>();
-        }
+    public IncrementalEstimator createIncrementalEstimator(ResolvableConjunction conjunction) {
+        if (!conjunctionModels.containsKey(conjunction)) buildConjunctionModel(conjunction);
+        return new IncrementalEstimator(conjunctionModels.get(conjunction));
+    }
 
-        private void add(ResolvableConjunction conjunction) {
-            stack.add(conjunction);
-            concludables.put(conjunction, null);
-        }
+    public static class IncrementalEstimator {
+        private final ConjunctionModel conjunctionModel;
+        private final Map<LocalModel, Pair<Double, Optional<Variable>>> modelScale;
```

Comment:
I think to be honest if we can get rid of the Optional we should - not having a scale for a model we can just reflect by having a `getOrDefault`?
It might even better to split this into two maps: Model -> Double for modelScale, and Model -> Variable for modelScaleSource ?